### PR TITLE
ISPN-13640 ResilienceMaxRetryIT is failing randomly

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.redhat.com/browse/ISPN-13640

Replacing https://github.com/infinispan/infinispan/pull/9814 and https://github.com/infinispan/infinispan/pull/9821

@diegolovison I opened my own PR in the end to add a high level description in the class javadoc and to also simplify the code a bit.

### CODE REVIEW

File: server/tests/src/test/java/org/infinispan/server/resilience/ResilienceMaxRetryIT.java [hunk 1]
```diff
@@ -34,78 +41,65 @@
...
+            cache.get(ThreadLocalRandom.current().nextInt());
+            break;
          } catch (TransportException e) {
-            // if you do 10 operations after you kill port1, the first 0 or 1 or 2 operations might fail.
-            // But if the 3rd operation is successful, then operations 4-10 should be successful as well
-            assert i == 0 || i == 1 || i == 2;
-            // assert that the failed server is the one that we killed
-            assert e.getMessage().contains(killedServer.toString());
+            // Assert that the failed server is the one that we killed
+            assert e.getMessage().contains(stoppedServerAddress.toString()) : e.getMessage();
```
Comment (pruivo) [inline server/tests/src/test/java/org/infinispan/server/resilience/ResilienceMaxRetryIT.java:67]: `assertTrue`?

File: server/tests/src/test/java/org/infinispan/server/resilience/ResilienceMaxRetryIT.java [hunk 2]
```diff
@@ -34,78 +41,65 @@
...
+      // Check that the stopped server was properly removed from the list
+      Collection<InetSocketAddress> currentServers = cache.getRemoteCacheManager().getChannelFactory().getServers(cacheNameBytes);
+      assertEquals(2, currentServers.size());
       for (InetSocketAddress currentServer : currentServers) {
-         if (currentServer.getHostString().equals(killedServer.getHostAddress())) {
-            throw new IllegalStateException("The removed server should not be present in the list");
+         if (currentServer.equals(stoppedServerAddress)) {
+            throw new IllegalStateException(stoppedServerAddress + " should not be present in the list: " + currentServers);
          }
       }
```
Comment (pruivo) [inline server/tests/src/test/java/org/infinispan/server/resilience/ResilienceMaxRetryIT.java:78]: `assertFalse(currentServers.contains(stoppedServerAddress));`?

File: server/tests/src/test/java/org/infinispan/server/resilience/ResilienceMaxRetryIT.java [hunk 3]
```diff
@@ -34,78 +41,65 @@
...
+      // so the number of operations we need depends on which server owns each key
       for (int i = 0; i < 10; i++) {
-         String random = UUID.randomUUID().toString();
          try {
-            cache.put(random, random);
-            itWorked = true;
+            cache.get(ThreadLocalRandom.current().nextInt());
             break;
          } catch (Exception e) {
             // the failure is expected
```
Comment (pruivo) [inline server/tests/src/test/java/org/infinispan/server/resilience/ResilienceMaxRetryIT.java:94]: check exception message?

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java [hunk 4]
```diff
@@ -138,11 +139,12 @@ public void prepare(String name) {
 
    @Override
    public void start(String name) {
+      this.name = name;
```
Comment (pruivo) [inline server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java:142]: shouldn't be set on `prepare` instead?

**Review by pruivo — CHANGES_REQUESTED** [review]

File: server/tests/src/test/java/org/infinispan/server/resilience/ResilienceMaxRetryIT.java (hunk 1 above)
Comment (danberindei) [reply server/tests/src/test/java/org/infinispan/server/resilience/ResilienceMaxRetryIT.java]: Done

File: server/tests/src/test/java/org/infinispan/server/resilience/ResilienceMaxRetryIT.java (hunk 2 above)
Comment (danberindei) [reply server/tests/src/test/java/org/infinispan/server/resilience/ResilienceMaxRetryIT.java]: I replaced it with a positive assert

File: server/tests/src/test/java/org/infinispan/server/resilience/ResilienceMaxRetryIT.java (hunk 3 above)
Comment (danberindei) [reply server/tests/src/test/java/org/infinispan/server/resilience/ResilienceMaxRetryIT.java]: Done

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java (hunk 4 above)
Comment (danberindei) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/AbstractInfinispanServerDriver.java]: Done

**Comment by danberindei** [comment]
Updated @pruivo!

I also made some changes to the driver so that clients are created with the same host string they'll receive from the servers instead of using `getHostName()` (which may transform 127.0.0.1 into localhost).

**Comment by pruivo** [comment]
@danberindei it looks like the authn/authz tests are failing now.

**Comment by danberindei** [comment]
I see what went wrong, I changed the "org.infinispan.test.host.address" system property to be `127.0.0.1` instead of `localhost` and that broke the Keycloak config. 
I reverted just that change for now, but if integration tests still fail I'll revert all the driver changes and change the test to compare "resolved" addresses instead of "unresolved" ones.

**Comment by danberindei** [comment]
CI looks good, finally!

**Review by pruivo — APPROVED** [review]

**Comment by pruivo** [comment]
merged! thanks @danberindei !

**Pull request merged**